### PR TITLE
Move foreign_types to overrideChildTca

### LINKED ISSUE #739: Deprecation: #80000 - InlineOverrideChildTca
Deprecation: #80000 - InlineOverrideChildTca

https://docs.typo3.org/c/typo3/cms-core/master/en-us/Changelog/8.7/Deprecation-80000-InlineOverrideChildTca.html

.. include:: ../../Includes.txt

============================================

Deprecation: #80000 - InlineOverrideChildTca

============================================

See :issue:`80000`

Description

===========

These :php:`TCA` :php:`type=inline` properties have been deprecated and superseded with the more

general property :php:`overrideChildTca`:

* foreign_types

* foreign_selector_fieldTcaOverride

* foreign_record_defaults

Impact

======

It is now possible to override display (FormEngine) related columns properties and the types section of

child :php:`TCA` from within the parent :php:`TCA`. This is also allowed in a

parents :php:`['types']['columnsOverrides']` section.

Affected Installations

======================

Instances using one of the above inline properties should adapt to the new :php:`overrideChildTca` property.

Migration

=========

A :php:`TCA` auto-migration is in place. It will transfer the old settings to the new property as

shown below and logs deprecation entries if there is no :php:`overrideChildTca` defined. This allows extension

authors to keep both the old and the new settings to support CMS v7 and v8 at the same time without

having deprecations logged.

foreign_types before and after transition to overrideChildTca:

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'foreign_types' => [

                    'aForeignType' => [

                        'showitem' => 'aChildField',

                    ],

                ],

                ...

            ],

        ],

        ...

    ],

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'overrideChildTca' => [

                    'types' => [

                        'aForeignType' => [

                            'showitem' => 'aChildField',

                        ],

                    ],

                ],

                ...

            ],

        ],

        ...

    ],

foreign_selector_fieldTcaOverride before and after transition to overrideChildTca:

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'foreign_selector' => 'uid_local',

                'foreign_selector_fieldTcaOverride' => [

                    'config' => [

                        'appearance' => [

                            'elementBrowserType' => 'file',

                        ],

                    ],

                ],

                ...

            ],

        ],

        ...

    ],

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'foreign_selector' => 'uid_local',

                'overrideChildTca' => [

                    'columns' => [

                        'uid_local' => [

                            'config' => [

                                'appearance' => [

                                    'elementBrowserType' => 'file',

                                ],

                            ],

                        ],

                    ],

                ],

                ...

            ],

        ],

        ...

    ],

foreign_record_defaults before and after transition to overrideChildTca:

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'foreign_record_defaults' => [

                    'aChildField' => 42,

                ],

                ...

            ],

        ],

        ...

    ],

.. code-block:: php

    'columns' => [

        'aField' => [

            'config' => [

                'type' => 'inline',

                'overrideChildTca' => [

                    'columns' => [

                        'aChildField' => [

                            'config' => [

                                'default' => 42,

                            ],

                        ],

                    ],

                ],

                ...

            ],

        ],

        ...

    ],

.. index:: Backend, TCA

### PR DESCRIPTION
**Pull request by MK-42**
fixes #739 

feel free to drop the last refactoring commit if you don't see its benefit. I'm not certain if the extraction of the generator helps, but it might come in handy for further refactorings

### CODE REVIEW

**Comment by MK-42** [comment]
I'll leave the merge conflict as is, but can certainly rebase this patch, or merge master.
whatever you prefer

File: stubs/Core/Utility/ExtensionManagementUtility.php [hunk 1]
```diff
@@ -47,4 +47,19 @@ public static function loadExtLocalconf(bool $false): void
     public static function extRelPath(string $key): void
     {
     }
+
```
Comment (sabbelasichon) [inline stubs/Core/Utility/ExtensionManagementUtility.php:50]: I would remove the docblocks here.

File: src/Rector/v8/v7/MoveForeignTypesToOverrideChildTcaRector.php [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            return null;
+        }
+
+        if (! $columns->value instanceof Array_) {
+            return null;
+        }
+
+        foreach ($this->extractColumnConfig($columns->value) as $columnConfig) {
+            //handle the special case of ExtensionManagementUtility::getFileFieldTCAConfig
+            if ($columnConfig instanceof StaticCall) {
```
Comment (sabbelasichon) [inline src/Rector/v8/v7/MoveForeignTypesToOverrideChildTcaRector.php:109]: Would create a dedicated method here for better readibility.

**Comment by sabbelasichon** [comment]
@MK-42. Really great. Well done. Do you think it is necessary to check for the type inline as a guard clause? Or are these configuration keys only used in this context?

File: stubs/Core/Utility/ExtensionManagementUtility.php (hunk 1 above)
Comment (MK-42) [reply stubs/Core/Utility/ExtensionManagementUtility.php:50]: I have slimmed them down to only the type-annotations, as not all parameters have native typehints yet

File: src/Rector/v8/v7/MoveForeignTypesToOverrideChildTcaRector.php (hunk 2 above)
Comment (MK-42) [reply src/Rector/v8/v7/MoveForeignTypesToOverrideChildTcaRector.php:109]: :+1:

**Comment by MK-42** [comment]
> @MK-42. Really great. Well done. Do you think it is necessary to check for the type inline as a guard clause? Or are these configuration keys only used in this context?

I thought about that, but the runtime-migration of TYPO3 does not check for that, but unconditionally moves the config over. I think it is reasonable to do the same here

**Comment by MK-42** [comment]
rebased on current master to resolve conflicts. should be ready for merge now

**Pull request merged**